core(graph): capture

romintomasetti opened the pull request:
## Summary

This PR introduces the *capture* node. Capture is only supported for the following backends:
- `Cuda`
- `HIP`
- `SYCL` (though in `SYCL` native language it's called this *recording*)

As the capture mechanism is backend-specific, the API has been prefixed with the backend.

It's been documented in:
- https://github.com/kokkos/kokkos-core-wiki/pull/652

### `Cuda` capture

The user passes a callable. We call it with a stream for which capture is enabled, and add the captured nodes as a sub-graph:
```c++
auto my_captured_node = predecessor.cuda_capture(
    exec,
    [data = data](const Kokkos::Cuda& exec_) { ... }
);
```

### `HIP` capture

The user passes a callable. We call it with a stream for which capture is enabled, and add the captured nodes as a sub-graph:
```c++
auto my_captured_node = predecessor.hip_capture(
    exec,
    [data = data](const Kokkos::HIP& exec_) { ... }
);
```

### `SYCL` recording

The user passes a callable. We call it with a queue for which recording was enabled, and add the recorded (executable) graph as a sub-graph:

```c++
auto my_captured_node = predecessor.sycl_capture(
    exec,
    [data = data](const Kokkos::SYCL& exec_) { ... }
);
```

<details>

<summary>

:warning: This is the old PR description. Kept for the record only. :older_man: 

</summary>

## Summary

This PR introduces the `then` node type for `Kokkos::Graph`.

The idea of the `then` node type is to allow users to add workloads that **are not defined through a `Kokkos` parallel construct** through the public API.

A formal definition could be:
> A `then` node places its workloads in a graph-ordered execution stack.
How workloads are actually defined and stored is backend-specific to best match the available backend execution models.

## Description

### What it does

The behavior of the `then` node varies:
* For `Cuda`, `HIP` and `SYCL`, a `then` node is equivalent to adding workloads through *graph capture* to the underlying native graph.
* For backends using the defaulted graph implementation, a `then` node stores the workload in a functor and calls this functor when calling `submit` (much like `Kokkos::Impl::ParallelFor` would do for instance).

In both cases, the users gives a functor and in both cases it is stored:
* For graph capturing backends, the functor is called once to capture workloads, and must be stored to keep any data alive.
* For others, it is obviously stored to be called on `submit`.

### Example

I've added a test in `core/unit_test/TestGraph.hpp` that should be the reference example. This is a code example:
```c++
auto node_then = other_node.then(
    [data = data] -> void (const Kokkos::Cuda& exec){
        my_external_kernel<<<..., ..., ..., exec.cuda_stream()>>>(data);
});
```

</details>

romintomasetti commented:
I wanted to put this PR in draft mode, but it seems I'm not allowed to.

maartenarnst commented on core/src/impl/Kokkos_GraphNodeImpl.hpp:
```diff
@@ -144,6 +144,11 @@ struct GraphNodeImpl<ExecutionSpace, Kernel,
+  template <class KernelDeduced>
+  GraphNodeImpl(ExecutionSpace const& ex, _graph_node_then_ctor_tag,
+                KernelDeduced&& arg_kernel)
+      : base_t(ex), m_kernel((KernelDeduced&&)arg_kernel) {}
```
```suggestion
      : base_t(ex), m_kernel(std::forward<KernelDeduced>(arg_kernel)) {}
```
(or keep it as is for consistency with the surrounding code?)

romintomasetti replied:
I think they will prefer consistency.

maartenarnst commented on core/src/Cuda/Kokkos_Cuda_Graph_Impl.hpp:
```diff
@@ -51,11 +51,21 @@ struct GraphImpl<Kokkos::Cuda> {
+  // keep a reference to 'then' nodes, so we need to guarantee that someone (us)
+  // stores the functor used for the capture, that might be keeping some objects
+  // alive.
+  std::vector<std::shared_ptr<erased_node_impl_t>> m_then_nodes;
```
The default implementation session the `GraphNodeBackendSpecificDetails` as the common type for the storage. It could be an option here too, ie., to use the `node_details_t` type.

- https://github.com/kokkos/kokkos/blob/9f7d134bd57a3e1fced0e7b799042a26317522d4/core/src/impl/Kokkos_Default_Graph_Impl.hpp#L53

romintomasetti replied:
Since `GraphNodeImpl` derives from `GraphNodeBackendSpecificDetails` https://github.com/kokkos/kokkos/blob/4889d4753daea6975d28589e662fa90825f6b88c/core/src/impl/Kokkos_GraphNodeImpl.hpp#L40-L44 we would probably store less if we keep `GraphNodeBackendSpecificDetails` (*e.g.* we would not store what `ExecutionSpaceInstanceStorage` stores).

romintomasetti replied:
Done. Let's see if it works.

romintomasetti commented on core/src/impl/Kokkos_Default_GraphNodeKernel.hpp:
```diff
@@ -95,6 +95,21 @@ struct GraphNodeAggregateKernelDefaultImpl
+struct GraphNodeThenImpl : public GraphNodeThenDefaultImpl<ExecutionSpace> {
+  Functor m_functor;
+
+  template <typename T>
+  explicit GraphNodeThenImpl(T &&functor)
+      : m_functor(std::forward<T>(functor)) {}
```
We discussed about it, and decided it would not be a good idea.
This proposed change was motivated by the default graph implementation. In this implementation, the kernel nodes created with a `Kokkos` parallel construct are executed on the execution space instance given to the policy during definition, such that the user can still "tweak" the default graph implementation to run nodes on different `space`s and effectively get asynchronicity (note that this is because kernel nodes of the default implementation derive from `Kokkos::Impl::Parallel...` structs). But we consider this a hacky hack since it should be the graph itself that decides where its nodes are executed.
So letting the user place a `then` node of the default graph implementation would imply that the following is allowed:
```c++
auto my_node_on_specified_exec = node.then(exec, ...);
```
but it's definitely not a good idea, especially for non-defaulted backends.

romintomasetti replied:
Outdated. A `then` node has been added in #7629.

maartenarnst commented on core/src/Cuda/Kokkos_Cuda_GraphNodeKernel.hpp:
```diff
@@ -34,6 +34,37 @@
+  // This explicit constructor is not needed by all compilers.
+  explicit GraphNodeThenImpl(Functor&& f) : m_functor(std::move(f)) {}
+
+  void capture(const Kokkos::Cuda& exec, cudaGraph_t& graph) {
+    // Set the underlying stream to capture mode.
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaStreamBeginCapture(
+        exec.cuda_stream(), cudaStreamCaptureModeGlobal));
```
We may need to think which exec/stream to use for the capture. Because we may need to avoid "side effects" where the user may be doing something else on the same exec/stream that might get captured too.

romintomasetti replied:
The rationale for now is that the stream used for capturing is the one from the policy.

maartenarnst commented on core/src/impl/Kokkos_GraphImpl.hpp:
```diff
@@ -31,6 +31,10 @@
 namespace Kokkos {
 namespace Impl {
 
+template <typename T>
+struct is_graph_then<T, std::enable_if_t<Kokkos::Impl::is_specialization_of_v<
+                            T, GraphNodeThenImpl>>> : public std::true_type {};
```
A related issue may be to think about a convention for organizing/naming/... new node types and how the concepts/constraints/... should behave.

Currently, we have the `GraphNodeImpl<ExecutionSpace, Kernel, Predecessor>` that holds the `m_kernel` member of type `Kernel`. And, currently, this `Kernel` is an instance of `GraphNodeKernelImpl<ExecutionSpace, PolicyType, Functor, PatternTag, ...>`. It can also be an instance of `GraphNodeAggregateKernelDefaultImpl<ExecutionSpace>` for the default implementation and of similar classes for non-default backends. The concept `Kokkos::Impl::is_graph_kernel` is `true` for all such instances, i.e., currently, it is `true` for aggregate nodes too.

With the proposed change to add the "then", we'll have:
- `GraphNodeKernelImpl<ExecutionSpace, PolicyType, Functor, PatternTag, ...>`
- `GraphNodeAggregateKernelDefaultImpl<ExecutionSpace>` and similar classes for non-default backends 
- `GraphNodeThenImpl<ExecutionSpace, Functor>`

And, with the proposed change, would not be `true` for the "then" node, if I understand correctly. The concepts `is_graph_kernel` and `is_graph_then` are also implemented differently, i.e., `is_graph_kernel` checks for the presence of a member type alias, and `is_graph_then` uses the `is_specialization_of_v`.

So, these are a few thoughts:
- How to get consistency in the naming? E.g. "Kernel" in the name of the aggregate node, but not in the name of the "then" node. One option may be to rename `GraphNodeAggregateKernelDefaultImpl` as `GraphNodeAggregateDefaultImpl`. With similar renaming for the non-default backends.
- What should `is_graph_kernel` do? Should it be true for an aggregate node? One option may be to define all such concepts like it is proposed now for the `is_graph_then`, i.e. with `is_specialization_of_v`. So `is_graph_kernel` would check whether the type is an instance of `GraphNodeKernelImpl`, and it would evaluate to false for the case of an aggregate node. Such a change would likely require some modifications to constraints/asserts in the code.

maartenarnst replied:
Done in extracted PRs.

crtrott commented:
I thought about this a bunch now and I think this is not a viable use of then. If I as a user create a graph and I do something like this:

```
graph.then([]() { printf("Hello\n"); );
```

I would absolutely expect that `graph.submit()` or `graph.execute()` does print "Hello". The concept of stuff inside `then` is captured partially by some backend dependent mechanism does not seem viable to me at all as a semantics for Kokkos graph. This sounds more like an inter-op thing if at all, something like `graph.cuda_capture(functor)`. Trying to make this backend agnostic feels like a fools errand to me because its gonna be so finicky what gets captured and what doesn't and you have to be aware of all kinds of backend specific stuff anyway.

romintomasetti commented:
> I thought about this a bunch now and I think this is not a viable use of then. If I as a user create a graph and I do something like this:
> 
> ```
> graph.then([]() { printf("Hello\n"); );
> ```
> 
> I would absolutely expect that `graph.submit()` or `graph.execute()` does print "Hello". The concept of stuff inside `then` is captured partially by some backend dependent mechanism does not seem viable to me at all as a semantics for Kokkos graph. This sounds more like an inter-op thing if at all, something like `graph.cuda_capture(functor)`. Trying to make this backend agnostic feels like a fools errand to me because its gonna be so finicky what gets captured and what doesn't and you have to be aware of all kinds of backend specific stuff anyway.

@dalg24 and I were also concerned about this. I think I went too quick on this PR, but in another repository wherein I drafted this, the user **had to give** an execution space type:
```c++
auto captured_node = pred.then<Kokkos::Cuda>(...);
```
Enforcing the type to be given would allow us to clearly document what happens for each backend. It would also help us prohibit a `then<Cuda>` for a graph on `HIP` and so on.

I still think it's worth having a method whose name is `then` whatever the backend. Consider this:
```c++
__device__ void my_function() {...}

int main() {
    using execution_space = Kokkos::DefaultExecutionSpace; // e.g. HIP or Cuda
    ...
    auto captured = pre.then<execution_space>([...](...){my_function(...);});
}
```

@crtrott Would enforcing the execution space type to be given change or mind? :smiley:

crtrott commented:
No adding a mandatory execution space argument doesn't convince me. The only thing I would accept is: graph.cuda_capture(f); etc.

romintomasetti commented:
OK, I think I got your point that this was not the right usage for `then`.

Going back to https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p2300r10.html#design-sender-adaptor-then, it's clear that `then` should encapsulate a function, not do some capture thingy.

Therefore, I think we might need 2 new node types:
- `then`
- `capture`

This PR would add the `capture` node. It would only work for `Cuda`, `HIP` and `SYCL`. Note that in the same way as we added a `Graph::native_graph` in #6904 for these 3 backends, I propose we don't make a backend-specific name and keep `capture`. In the end, these 3 backends work in a similar manner.

A followup PR would add the `then` node. I can provide some use cases we already have.

But I think this pseudo-code better shows what's the difference between a `then` and a `capture`:
```c++
// This would possibly capture/record many kernels (p-for, p-scan, p-reduce, TPLs, whatever).
// Many nodes can be added under the hood (it's a child graph node).
auto node_capture = pred.capture([data = data](const Kokkos::Cuda& exec) {... to be captured ...});

// This would add a single node, and would just call the lambda through a driver.
auto node_then = pred.then(KOKKOS_LAMBDA{ ... like a p-for from 0 to 1 ...});
```

Would you be OK if we revise this PR to limit it to adding the `capture` functionality to the 3 aforementioned backends ?

crtrott commented:
I still am firmly believing this needs a backend specific name, since what gets captured is backend specific. It also may go completely haywire. Like you do "cuda" capture, and call a function inside which calls Kokkos, which calls some Cuda kernel, which will get captured, but will not work upon submit because the scratch pad for reduction was reallocated in the meantime. This capturing stuff is a big foot gun, and it needs to be very clear that it is one. Naming stuff backend specific makes it clear something outside of normal Kokkos portability happens.

romintomasetti commented:
@crtrott I agree with you. I've reworked a bit the PR. For now, it's introducing 3 things and will need to be split in smaller chunks but it gives you (hopefully) a better overview. The 3 things:
- `then`
- `<backend>_capture`
- build a `Kokkos::Graph` from a native pointer

In the wings, we are building a concrete example usage of `Kokkos::Graph` that would benefit from these 3 features.

### `then`

The idea is that this `then` now means the same thing as the `then` in P2300. In essence, you pass a functor to the `then` and it's guaranteed to be executed at each graph submission, in the graph topology ordering.
```c++
auto my_then = predecessor.then([data = data]{... do some stuff with data ...});
```
You will note that the functor passed to the `then` must be callable without any argument. The functor will be stored by the `then` node to ensure that the `data`'s lifetime is bound to the graph's lifetime.

The `then` is therefore nothing else then a kernel launch (done through one of the `Kokkos` `__global__` drivers), and it adds exactly one node in the graph (one node in the `Kokkos::Graph` - but also one node in the backend graph).

For now, I've implemented it as a parallel-for from 0 to 1, but I think we can do better than that.

### `<backend>_capture`

Following your advice, the capture is now prefixed with the backend to emphasize that it's backend-specific, interoperability thing.
```c++
auto my_capture = predecessor.<backend>_capture(exec, [data = data](const Kokkos::Cuda& exec_){ ... some kernel launches that work on data ...});
```

In this case, `Kokkos` will call the functor in-between a start-stop capture on `exec` rightaway. The capture will create one node in the `Kokkos::Graph`, that will be responsible of keeping the functor alive. But `Kokkos` does not know how many nodes where added to the backend graph, because the capture mechanism is essentially returning a graph, that we add to the `Kokkos::Graph` as a "child graph node".

### Build a `Kokkos::Graph` from a native graph

This use case might have as many motivations as there are projects out there :smile: But mine is quite clear. If you use a `Cuda` conditional node, this node returns a handle to the graph that will be conditionally executed. The idea is that we give this handle to `Kokkos::Graph` to start adding nodes directly into this "conditional graph".

romintomasetti commented on core/unit_test/TestGraph.hpp:
```diff
@@ -813,18 +831,36 @@ struct GraphNodeTypes {
 constexpr bool test_is_graph_kernel() {
-  using types = GraphNodeTypes<TEST_EXECSPACE>;
-  static_assert(Kokkos::Impl::is_graph_kernel_v<types::kernel_t>);
-  static_assert(!Kokkos::Impl::is_graph_kernel_v<types::aggregate_t>);
-  static_assert(Kokkos::Impl::is_graph_kernel_v<types::then_t>,
+  using types = GraphNodeTypes<Exec>;
+  static_assert(Kokkos::Impl::is_graph_kernel_v<typename types::kernel_t>);
+  static_assert(!Kokkos::Impl::is_graph_kernel_v<typename types::aggregate_t>);
+  static_assert(Kokkos::Impl::is_graph_kernel_v<typename types::then_t>,
                 "This should be verified until the 'then' has its own path to "
                 "the driver.");
+  if constexpr (types::support_capture)
+    static_assert(!Kokkos::Impl::is_graph_kernel_v<typename types::capture_t>);
+  return true;
+}
+static_assert(test_is_graph_kernel<TEST_EXECSPACE>());
```
Note for the reviewers.

These changes are needed because when capture neither `Cuda`, `HIP` or `SYCL` is defined, there is no definition of `GraphNodeCaptureImpl`.

Since `is_graph_kernel_v` ultimately uses `std::is_base_of_v`, I was getting the error:
>  implicit instantiation of undefined template 'GraphNodeCaptureImpl'

when evaluating `is_graph_kernel_v<typename types::capture_t>`. This is because [of the following rule](https://en.cppreference.com/w/cpp/language/if#Constexpr_if):
> Outside a template, a discarded statement is fully checked.

So I had to make `test_is_graph_kernel` a templated function.

maartenarnst commented on core/src/Cuda/Kokkos_Cuda_GraphNodeKernel.hpp:
```diff
@@ -34,6 +34,35 @@
 namespace Kokkos {
 namespace Impl {
 
+template <typename Functor>
+struct GraphNodeCaptureImpl<Kokkos::Cuda, Functor> {
```
Should this be moved to separate file? Such as `core/src/Cuda/Kokkos_Cuda_GraphNodeCaptureImpl.hpp`? (Just like we had e.g. `Kokkos_GraphNodeThenImpl.hpp`)

romintomasetti replied:
I'm not sure it's worth adding 3 new files (one for each backend). It's true that the file's name has `NodeKernel` in it. I would wait for other opinions.

maartenarnst commented on core/unit_test/TestGraph.hpp:
```diff
@@ -864,6 +900,118 @@ TEST(TEST_CATEGORY, when_all_type) {
+        <<<dim3(1, 1, 1), dim3(1, 1, 1), 0, exec.cuda_stream()>>>(data.data());
+  }
+
+  template <typename N, typename T>
+  static auto add(const N& node, const Kokkos::Cuda& exec, const T& data) {
+    return node.cuda_capture(exec, [data = data](const Kokkos::Cuda& exec_) {
+      ExternalCapture::compute(exec_, data);
+    });
+  }
```
The first argument could also be named `predecessor` or `pred`. Just to make it clear that the result of the capture will be used to create a new node that will follow what we're passing as argument to this function.

romintomasetti replied:
Done :heavy_check_mark:

maartenarnst commented on core/unit_test/TestGraph.hpp:
```diff
@@ -864,6 +900,118 @@ TEST(TEST_CATEGORY, when_all_type) {
+#endif
+
+template <typename Exec, typename ViewType>
+struct ExternalCapture {
```
Another name could be `BackendCapture`. It's fine to keep it as is too.

romintomasetti replied:
I chose `External` because capture would be used to integrate "external" libraries into a graph.

maartenarnst commented on core/unit_test/TestGraph.hpp:
```diff
@@ -864,6 +900,118 @@ TEST(TEST_CATEGORY, when_all_type) {
+        SetViewToValueFunctor<TEST_EXECSPACE, int>{data, offset});
+
+    // We don't keep a reference to the created external node, to mimic that
+    // someone did use capture in some deep-down library call (e.g. in
```
```suggestion
    // someone used capture in some deep-down library call (e.g. in
```

romintomasetti replied:
Done :heavy_check_mark:

maartenarnst commented on core/unit_test/TestGraph.hpp:
```diff
@@ -801,6 +809,16 @@ struct GraphNodeTypes {
                                          Kokkos::Experimental::TypeErasedTag,
                                          Kokkos::Experimental::TypeErasedTag>;
 
+#if defined(KOKKOS_ENABLE_CUDA)
+  static constexpr bool support_capture = std::is_same_v<Exec, Kokkos::Cuda>;
+#elif defined(KOKKOS_ENABLE_HIP) && defined(KOKKOS_IMPL_HIP_NATIVE_GRAPH)
+    static constexpr bool support_capture = std::is_same_v<Exec, Kokkos::HIP>;
+#elif defined(KOKKOS_ENABLE_SYCL) && defined(SYCL_EXT_ONEAPI_GRAPH)
+  static constexpr bool support_capture = std::is_same_v<Exec, Kokkos::SYCL>;
+#else
+  static constexpr bool support_capture = false;
+#endif
```
Just wondering if we'd want to make this a trait of `Kokkos::Graph<Exec>`?

romintomasetti replied:
I also thought about doing it. Waiting for other opinions though.

romintomasetti commented on core/src/SYCL/Kokkos_SYCL_GraphNodeKernel.hpp:
```diff
@@ -30,6 +30,23 @@
+      const Kokkos::SYCL& exec,
+      sycl::ext::oneapi::experimental::command_graph<
+          sycl::ext::oneapi::experimental::graph_state::modifiable>& graph) {
+    // This seems wrong if we already have nodes in the graph.
```
Check this before we merge.

romintomasetti replied:
Done. I've tested it on the `Headroom` machine.

romintomasetti commented:
Retest this please.

romintomasetti commented:
@crtrott @dalg24 It works on `Cuda`, `HIP` and `SYCL` now. Would you mind having a first look? Thanks !

romintomasetti commented:
The build failure https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-7552/61/pipeline-console/?selected-node=41 is unrelated. Retest this please.

maartenarnst commented on core/src/Cuda/Kokkos_Cuda_GraphNodeKernel.hpp:
```diff
@@ -34,6 +34,35 @@
+  Functor m_functor;
+  cudaGraphNode_t m_node = nullptr;
+
+  void capture(const Kokkos::Cuda& exec, cudaGraph_t& graph) {
```
```suggestion
  void capture(const Kokkos::Cuda& exec, cudaGraph_t graph) {
```

It appears the `graph` is passed by value in other cases such as

https://github.com/kokkos/kokkos/blob/8cd597c7a7119f5984415db2b6f69d0ca38825b0/core/src/Cuda/Kokkos_Cuda_Instance.hpp#L281-L285

romintomasetti replied:
I makes sense since it's an opaque pointer. Done :heavy_check_mark:

maartenarnst commented on core/src/Kokkos_GraphNode.hpp:
```diff
@@ -246,6 +247,51 @@ class GraphNodeRef {
+      // Add the node itself to the backend's graph data structure, now that
+      // everything is set up.
+      graph_ptr->add_node(exec, rv.m_node_impl);
+      // Add the predecessaor we stored in the constructor above in the
+      //  backend's data structure, now that everything is set up.
```
```suggestion
      // Add the predecessaor we stored in the constructor above in the
      // backend's data structure, now that everything is set up.
```

romintomasetti replied:
:heavy_check_mark:

maartenarnst commented on core/src/Kokkos_GraphNode.hpp:
```diff
@@ -246,6 +247,51 @@ class GraphNodeRef {
+#if defined(KOKKOS_ENABLE_CUDA) ||                                           \
+    (defined(KOKKOS_ENABLE_HIP) && defined(KOKKOS_IMPL_HIP_NATIVE_GRAPH)) || \
+    (defined(KOKKOS_ENABLE_SYCL) && defined(SYCL_EXT_ONEAPI_GRAPH))
+  /// The parameter @p exec can be used for multi-GPU graphs.
+  /// The captured nodes will be assigned to the device associated with @p exec.
+  /// @todo Support label.
```
Should we add a few sentences here to say what this function does? E.g. 

> Capture the backend operations that calling the passed function object with the passed execution space instance enqueues into the underlying stream.

Also, should we add "warnings" to say that it's the user’s responsibility to ensure that no fences and no other backend operations are enqueued during the capture? And say a word about how the functor is stored in the node to ensure the lifetime of the needed data?

romintomasetti replied:
I've updated the wording a bit.

romintomasetti replied:
Done, OK for you ?

maartenarnst commented on core/src/Kokkos_GraphNode.hpp:
```diff
@@ -246,6 +247,51 @@ class GraphNodeRef {
+  /// The parameter @p exec can be used for multi-GPU graphs.
+  /// The captured nodes will be assigned to the device associated with @p exec.
+  /// @todo Support label.
+  template <class Functor>
```
Should we constrain the `Functor` like we had done for the "then" node? E.g. something like

```
std::enable_if_t<std::is_invocable_r_v<
    void, const Kokkos::Impl::remove_cvref_t<Functor>, const ExecutionSpace&>>>
```

romintomasetti replied:
:heavy_check_mark:

maartenarnst commented on core/unit_test/TestGraph.hpp:
```diff
@@ -866,6 +902,163 @@ TEST(TEST_CATEGORY, when_all_type) {
+
+template <typename Exec>
+struct ExternalCapture {
+  template <typename N, typename E, typename T, typename... Args>
+  static auto add(const N&, const E&, const T&, const T&, Args&&...) {
+    Kokkos::abort("Not implemented.");
+    return Kokkos::Experimental::GraphNodeRef<E>();
+  }
```
Just wondering if we really need this version? Because we're doing `GTEST_SKIP` in the cases in which we wouldn't end up in the version below?

maartenarnst commented on core/unit_test/TestGraph.hpp:
```diff
@@ -866,6 +902,163 @@ TEST(TEST_CATEGORY, when_all_type) {
+
+// clang-format off
+#define EXTERNAL_CAPTURE_ADD(__capture__, __backend__)                                                  \
+  template <typename N, typename DstType, typename... Srcs>                                             \
```
```suggestion
  template <typename N, typename DstType, typename... SrcTypes>                                             \
```
(or just `T` and `Args`)

romintomasetti replied:
Done for `SrcTypes` :heavy_check_mark:

maartenarnst commented on core/unit_test/TestGraph.hpp:
```diff
@@ -866,6 +902,163 @@ TEST(TEST_CATEGORY, when_all_type) {
+template <typename Exec>
+std::enable_if_t<GraphNodeTypes<Exec>::support_capture == false, void>
+test_graph_capture() {
+  GTEST_SKIP() << "The execution space " << Kokkos::Impl::TypeInfo<Exec>::name()
+               << " does not support capture.";
```
```suggestion
  GTEST_SKIP() << "The Graph backend for" << Kokkos::Impl::TypeInfo<Exec>::name()
               << " does not support capture.";
```

romintomasetti replied:
:heavy_check_mark:

maartenarnst commented on core/unit_test/TestGraph.hpp:
```diff
@@ -866,6 +902,163 @@ TEST(TEST_CATEGORY, when_all_type) {
+template <typename Exec>
+std::enable_if_t<GraphNodeTypes<Exec>::support_capture == true, void>
+test_graph_capture() {
+  using view_r1_t = Kokkos::View<int[5], Exec>;
+  using view_r0_t = Kokkos::View<int, Exec>;
```
```suggestion
  using view_t = Kokkos::View<int[5], Exec>;
```

romintomasetti replied:
I won't do it :wink:

maartenarnst commented on core/unit_test/TestGraph.hpp:
```diff
@@ -866,6 +902,163 @@ TEST(TEST_CATEGORY, when_all_type) {
+  exec.fence("Wait for data to be initialized.");
+
+  // clang-format off
+#define TAKE_R0_AT(index) const view_r0_t data_##index(Kokkos::subview(data, index));
+
+  TAKE_R0_AT(0) TAKE_R0_AT(1) TAKE_R0_AT(2) TAKE_R0_AT(3) TAKE_R0_AT(4)
```
```suggestion
#define TAKE_AT(index) const auto data_at_##index(Kokkos::subview(data, index));

  TAKE_AT(0) TAKE_AT(1) TAKE_AT(2) TAKE_AT(3) TAKE_AT(4)
```

romintomasetti replied:
:heavy_check_mark:

maartenarnst commented on core/unit_test/TestGraph.hpp:
```diff
@@ -866,6 +902,163 @@ TEST(TEST_CATEGORY, when_all_type) {
+  auto root  = Kokkos::Impl::GraphAccess::create_root_ref(graph);
+  // clang-format on
+
+  auto memset_left = root.then_parallel_for(
+      Kokkos::RangePolicy<Exec>(exec_left, 0, 1),
+      SetViewToValueFunctor<Exec, int>{data_0, offset_left});
+
+  auto memset_right = root.then_parallel_for(
+      Kokkos::RangePolicy<Exec>(exec_right, 0, 1),
+      SetViewToValueFunctor<Exec, int>{data_4, offset_right});
```
Is it right to say that because we're running this test on device backends only, the execution space instance that we pass here ultimately is not used? If that's right, should we remove the `exec_left` and `exec_right`? And could we have e.g. an `exec` to submit the graph on, and and `exec_capture` that we use for the capture?

maartenarnst replied:
@romintomasetti and I discussed about it. On a backend with a native graph implementation, the current impact of the `exec` passed to `then_parallel_for` is:
- With the constant and global launch mechanisms, the passed `exec` is used to copy the driver to device. With the local launch mechanism, there is no such use.
- In particular on `Cuda`, multi GPU graphs are supported. For such use cases, the passed `exec` can determine the particular device on which the graph node will be executed.   

We decided it makes sense to keep the `exec_left` and `exec_right` arguments because it's a general way of coding that is agnostic of which launch mechanism will be used. And it anticipates multi GPU use cases.

maartenarnst commented on core/unit_test/TestGraph.hpp:
```diff
@@ -866,6 +902,163 @@ TEST(TEST_CATEGORY, when_all_type) {
+  ASSERT_TRUE(contains(exec_right, data_4, 0));
+
+  // The view is shared by:
+  //  - this scope
+  //  - the memset nodes
+  //  - the capture nodes
```
```suggestion
  //  - this scope (1 + 5)
  //  - the memset nodes (2)
  //  - the capture nodes (2 + 2 + 3)
```

romintomasetti replied:
:heavy_check_mark:

maartenarnst commented on core/unit_test/TestGraph.hpp:
```diff
@@ -866,6 +902,162 @@ TEST(TEST_CATEGORY, when_all_type) {
+  auto captured_right =
+      ExternalCapture<Exec>::add(memset_right, exec_left, data_3, data_4);
+
+  // We don't keep a reference to the created external node, to mimic that
```
```suggestion
  // We don't keep a reference to the created external nodes, to mimic that
```

romintomasetti replied:
:heavy_check_mark:

maartenarnst commented:
Hi @dalg24 and @crtrott. Romin and I have iterated on this PR. Would you have some time to take a look?

Two particular open questions we have are:
- When we introduced the `then` node, we put it in a separate file. Here, we're introducing the `capture` node in existing files. Is this OK for you? 

https://github.com/kokkos/kokkos/pull/7552/files#r1930732246

- The test in this PR adds a property `support_capture` to detect whether the graph for a particular backend supports capture (i.e. whether it is Cuda, HIP or SYCL). Should this trait become a static member variable in Graph<Exec>?

https://github.com/kokkos/kokkos/pull/7552/files#r1930820701

Thanks in advance! Let's keep moving :)

dalg24 reviewed (commented):
Please draft the API documentation for it.

romintomasetti commented on core/unit_test/TestGraph.hpp:
```diff
@@ -867,6 +903,157 @@ TEST(TEST_CATEGORY, when_all_type) {
+                           3 * static_cast<int>(GraphNodeType::CAPTURE)));
+}
+
+TEST(TEST_CATEGORY, graph_capture) { test_graph_capture<TEST_EXECSPACE>(); }
```
I think I went with this for a good reason. I'll try to remember why :wink:

romintomasetti replied:
OK, I think I remember now. When you use `TEST(..., ...)`, it's defining the body of a "regular" function, *i.e.* it's not a templated one.

Therefore, the compiler will still parse the code for both branches of the `if constexpr`, and for backends that do not support capture, it would end up in compilation errors like for https://github.com/kokkos/kokkos/pull/7552#discussion_r2018935629.

dalg24 replied:
It does not sound right.  Please investigate.

romintomasetti replied:
Indeed, I'm investigating. CI ongoing :construction:

romintomasetti replied:
Jenkins failures seem unrelated, so indeed it wasn't needed :wink:

romintomasetti commented on core/src/Cuda/Kokkos_Cuda_Graph_Impl.hpp:
```diff
@@ -126,7 +126,7 @@ struct GraphImpl<Kokkos::Cuda> {
   std::enable_if_t<
       Kokkos::Impl::is_graph_kernel_v<typename NodeImpl::kernel_type>>
-  add_node(std::shared_ptr<NodeImpl> const& arg_node_ptr) {
+  add_node(std::shared_ptr<NodeImpl> arg_node_ptr) {
```
:+1:

romintomasetti commented on core/unit_test/TestGraph.hpp:
```diff
@@ -867,6 +903,157 @@ TEST(TEST_CATEGORY, when_all_type) {
 }
 
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP)
+template <GraphNodeType value, typename T, typename... Srcs>
```
I've renamed `T` to `DstType` and `Srcs` to `SrcTypes` to make them consistent with the rest of the test file. Thanks !

romintomasetti commented on core/unit_test/TestGraph.hpp:
```diff
@@ -867,6 +903,157 @@ TEST(TEST_CATEGORY, when_all_type) {
+struct ExternalCapture;
+
+// clang-format off
+#define EXTERNAL_CAPTURE_ADD(__capture__, __backend__)                                                      \
```
* OK for renaming to `KOKKOS_TEST_GRAPH_CAPTURE_DEFINE_ADD`.
* OK for not using the `__`.

Done :heavy_check_mark:

romintomasetti commented on core/unit_test/TestGraph.hpp:
```diff
@@ -867,6 +903,157 @@ TEST(TEST_CATEGORY, when_all_type) {
+#define EXTERNAL_CAPTURE_ADD(__capture__, __backend__)                                                      \
+  template <typename N, typename DstType, typename... SrcTypes>                                             \
+  static auto add(const N& pred, const Kokkos::__backend__& exec, const DstType& dst, SrcTypes&&... srcs) { \
+    /* TODO Use C++ 20 "Allow pack expansion in lambda init-capture" feature. */                            \
```
OK, and done :heavy_check_mark:

romintomasetti replied:
I would suggest not to do it and keep a single version of this macro. Agreed ?

dalg24 replied:
I won't block on this but keep in mind that C++20 will be required in 5.0 and that will come to develop sometime in the summer.  So you are trading a TODO that is likely to linger for a `#ifdef KOKKOS_ENABLE_CXX17` branch that will be found and removed.

romintomasetti replied:
I'll do it then :wink:

romintomasetti replied:
Done :heavy_check_mark:

romintomasetti commented on core/unit_test/TestGraph.hpp:
```diff
@@ -867,6 +903,157 @@ TEST(TEST_CATEGORY, when_all_type) {
+  EXTERNAL_CAPTURE_ADD(sycl_capture, SYCL)
+};
+#endif
+
```
OK, and done :heavy_check_mark:

dalg24 commented on core/unit_test/TestGraph.hpp:
```diff
@@ -816,18 +834,36 @@ struct GraphNodeTypes {
+      Kokkos::Impl::GraphNodeCaptureImpl<Exec, CountTestFunctor<Exec>>;
 };
 
+template <typename Exec>
```
I think it is a sign of a defect in our internals, in the way we detect whether capture is available or not.
Providing an empty definition for `Impl::GraphNodeCaptureImpl` would get this to compile but I haven't thought through whether that is the right course for action.

romintomasetti replied:
Let's recap.

The following does not work because the compiler still parses the `if constexpr` branch, and so it tries to instantiate `Impl::GraphNodeCaptureImpl` for `Kokkos::Serial`, which we don't support for capture; so there is no definition for it and it ends up in a compile error because the `is_graph_kernel_v` concept (based on `KOKKOS_IMPL_IS_CONCEPT`) uses `is_base_of_v`.
```c++
constexpr bool test_is_graph_kernel() {
  using types = GraphNodeTypes<TEST_EXECSPACE>;
  if constexpr (types::support_capture)
    static_assert(!Kokkos::Impl::is_graph_kernel_v<typename types::capture_t>);
  return true;
}
static_assert(test_is_graph_kernel());
```

The following code is the same, except that `test_is_graph_kernel` is now a template, such that the line within the `if constexpr` is not parsed and we don't have the compile error we'd get from the code above. So either we:
1. keep `Impl::GraphNodeCaptureImpl` with no definition and keep `test_is_graph_kernel` templated
2. or we provide an empty definition for `Impl::GraphNodeCaptureImpl` and `test_is_graph_kernel` can stay un-templated.

```c++
template <typename Exec>
constexpr bool test_is_graph_kernel() {
  using types = GraphNodeTypes<Exec>;
  if constexpr (types::support_capture)
    static_assert(!Kokkos::Impl::is_graph_kernel_v<typename types::capture_t>);
  return true;
}
static_assert(test_is_graph_kernel<TEST_EXECSPACE>());
```

Now, the reason why we don't have such issues with the `test_is_graph_capture` test is because the underlying tested concept, `is_graph_capture_v`, does not use `KOKKOS_IMPL_IS_CONCEPT` macro, so it does stay away from using `is_base_of_v`. So even if the `if constexpr` branch is parsed by the compiler, it's still fine.

I would prefer keeping `Impl::GraphNodeCaptureImpl` not defined for backends that do not support capture natively, but I'd be fine either way.

dalg24 replied:
Fine

dalg24 commented:
> Please draft the API documentation for it.

I meant it

romintomasetti commented on core/src/Kokkos_GraphNode.hpp:
```diff
@@ -246,6 +247,63 @@ class GraphNodeRef {
+#if defined(KOKKOS_ENABLE_CUDA) ||                                           \
+    (defined(KOKKOS_ENABLE_HIP) && defined(KOKKOS_IMPL_HIP_NATIVE_GRAPH)) || \
+    (defined(KOKKOS_ENABLE_SYCL) && defined(SYCL_EXT_ONEAPI_GRAPH))
+  /// Capture the backend operations that calling the passed function object
+  /// with the passed execution space instance enqueues into the underlying
+  /// backend queue.
+  /// @note The function object will be stored by the underlying graph
+  ///       implementation, thereby ensuring that any captured data is kept
+  ///       alive until the graph is destroyed.
+  /// @note The parameter @p exec can be used for multi-GPU graphs.
+  ///       The captured nodes will be assigned to the device associated with
+  ///       @p exec.
+  /// @warning It's the user's responsability to ensure that theere is no
+  ///          illegal operation going on while capturing.
+  /// @todo Support label.
```
@dalg24 Is it sufficient for a draft ?

romintomasetti replied:
So with this *docstring* and the example I just pushed, would it be OK ? (@dalg24 @maartenarnst)

romintomasetti replied:
Let's get started with:
- https://github.com/kokkos/kokkos-core-wiki/pull/648

romintomasetti replied:
And the next one is:
- https://github.com/kokkos/kokkos-core-wiki/pull/652

romintomasetti commented:
@dalg24 What's the step forward? :smile:

romintomasetti commented:
@crtrott Build failure are either unrelated, or I'm missing the `SNL` approval. Would you have some time for a quick feedback on this ? Thanks !

dalg24 commented:
Retest this please

romintomasetti commented on core/unit_test/TestGraph.hpp:
```diff
@@ -867,6 +903,168 @@ TEST(TEST_CATEGORY, when_all_type) {
+    test_graph_capture<TEST_EXECSPACE>();
+  } else {
+    GTEST_SKIP() << "The graph backend for "
+                 << Kokkos::Impl::TypeInfo<TEST_EXECSPACE>::name()
```
Forgot about it, and done :heavy_check_mark:

romintomasetti commented on core/src/Kokkos_GraphNode.hpp:
```diff
@@ -246,6 +247,51 @@ class GraphNodeRef {
+      // Add the node itself to the backend's graph data structure, now that
+      // everything is set up.
+      graph_ptr->add_node(exec, rv.m_node_impl);
+      // Add the predecessaor we stored in the constructor above in the
```
Done for all occurrences :+1:

romintomasetti commented on core/unit_test/TestGraph.hpp:
```diff
@@ -867,6 +903,168 @@ TEST(TEST_CATEGORY, when_all_type) {
+
+template <typename Exec>
+void test_graph_capture() {
+  using view_t = Kokkos::View<int[5], Exec>;
```
Indeed, using a `using` once in this case isn't helping the reader. Done :heavy_check_mark:

romintomasetti commented on core/unit_test/TestGraph.hpp:
```diff
@@ -867,6 +903,168 @@ TEST(TEST_CATEGORY, when_all_type) {
+               offset_right + static_cast<int>(GraphNodeType::CAPTURE)));
+  ASSERT_TRUE(contains(exec_graph, data_2,
+                       offset_left + offset_right +
+                           3 * static_cast<int>(GraphNodeType::CAPTURE)));
```
Done, please don't judge my art too much :art:

romintomasetti commented:
Build failures seem unrelated.

crtrott commented on core/src/Kokkos_GraphNode.hpp:
```diff
@@ -246,6 +247,51 @@ class GraphNodeRef {
+      graph_ptr->add_predecessor(rv.m_node_impl, *this);
+      KOKKOS_ENSURES(bool(rv.m_node_impl))
+      return rv;
+    }
```
How about this: 
```suggestion
    } else {
       Kokkos::abort("Some appropriate message");
    }
```

crtrott replied:
Or actually how about replacing the if constexrp above with `static_assert()`

romintomasetti replied:
It tried to be consistent with what we did in #6904:
https://github.com/kokkos/kokkos/blob/ca8d31a245bc553b4de6d48e62ce8efe0c336154/core/src/Kokkos_Graph.hpp#L185-L201

And to quote @dalg24 at the time (https://github.com/kokkos/kokkos/pull/6904#discussion_r1767996063):
> They'll get a compile error if you don't return anything.

crtrott reviewed (commented):
Agree that the test feels a bit overcomplicated, but it's fine. Took me a moment to follow what's going on though. Here is a suggestion: your macro thing and partially specialized classes. Just replace it with a single class, use Kokkos::DefaultExecutionSpace, use ifdef KOKKOS_ENABLE_CUDA/HIP/SYCL inside the compute function body (the signature is now common since it uses Kokkos::DefaultExecutionSpace) and use an ifdef around the capture name in the add function (whose signature is now also common). 

It's probably less code, no multiline macro anymore and the ifdefs are pretty local too.

romintomasetti commented:
I've reworked the test to make it easier to read (less fancy stuff).

maartenarnst commented on core/unit_test/TestGraph.hpp:
```diff
@@ -867,6 +903,160 @@ TEST(TEST_CATEGORY, when_all_type) {
+    sycl_capture
+#endif
+    (exec, [dst, tup = std::make_tuple(std::forward<SrcTypes>(srcs)...)](const Exec& exec_) {   \
+            std::apply([&](const auto&... args) {                                                         \
+                ExternalCapture::compute(exec_, dst.data(), args.data()...);}, tup);                      \
+    });                                                                                                   \
```
The alignment of the `\` seems a bit off?

romintomasetti replied:
Indeed, and done :heavy_check_mark:

crtrott reviewed (approved):
Yeah the test is better, easier to follow thanks.

romintomasetti commented:
@dalg24 Would you have some time to give this one the final ✅? I quickly looked at the failing jobs, it seems the failures are unrelated. Thanks!

dalg24 commented:
Please add it to the 4.7 changelog

romintomasetti commented:
> Please add it to the 4.7 changelog

Done :heavy_check_mark:
